src/main/java/com/chessmates/utility/LichessResultSet.groovy

```diff
@@ -0,0 +1,75 @@
+package com.chessmates.utility
+
+import java.util.function.Function
+import java.util.function.Predicate
+
+/**
+ * Class responsible for fetching a full lichess result set from multiple pages.
+ */
+class LichessResultSet<T> {
+
+    interface GetPageFunction<T> extends Function<Integer, LichessResultPage<T>> {}
+
+    private GetPageFunction<T> getPage
+    private Predicate<T> shouldStop
+
+
+    /**
+     * The result set must be constructed with a set of closures defining certain aspects of its behaviour.
+     *
+     * @param getPage Get a new results page for a given page number.
+     * @param shouldStop Called for every result in a page - return true to stop, otherwise result set will stop building all pages fetched.
+     */
+    LichessResultSet(GetPageFunction<T> getPage, Predicate<T> shouldStop) {
+        this.getPage = getPage
+        this.shouldStop = shouldStop
+    }
+
+    /**
+     * Get all pages for the given result set.
+     */
+    List<T> get() {
+        def items = []
+
+        // Start at page one and keep requesting pages until there are no more pages.
+        def previousPage = getPage.apply(1)
+        items.addAll(getItemsFromPage(previousPage, shouldStop)['items'])
+
+        // Loop around all following pages until the stop condition is fulfilled or we hit the end of the results.
+        while(previousPage?.nextPage != null) {
```

Comment:
For the same reason as below I'm going to leave it an explicit null check. This could conceivably work with negative page numbers, and in that case 0 would be a valid option.